### PR TITLE
fix(pattern): remove non-required properties from PatternObject and fix some type issues.

### PR DESCRIPTION
**Pull request by plainheart**
fixes apache/echarts#14953

1) make `type`, `svgElement`, `svgWidth`, `svgHeight` of `PatternObject ` optional.
2) fixed some type issues.

### CODE REVIEW

File: src/graphic/Pattern.ts [hunk 1]
```diff
@@ -5,16 +5,16 @@ type CanvasPatternRepeat = 'repeat' | 'repeat-x' | 'repeat-y' | 'no-repeat'
 export interface PatternObject {
     id?: number
 
-    type: 'pattern'
+    type?: 'pattern'
 
     image: ImageLike | string
```
Comment (pissang) [inline src/graphic/Pattern.ts:10]: Is it better to use `union of interface`?
```ts
interface CanvasPatternObject {
  type: 'pattern'
  image: ImageLike | string
  repeat: ...
}
interface SVGPattenObject  {
  type: 'pattern'
  svgElement?: SVGElement
  ....
}

type PatternObject = CanvasPatternObject | SVGPatternObject
```

Comment (plainheart) [reply src/graphic/Pattern.ts:10]: I'm not sure which way is better,

1) Use union types (you mentioned above)
2) Define a parent type named `PatternObject`, `CanvasPatternObject ` and `SVGPatternObject` extend it.

Comment (pissang) [reply src/graphic/Pattern.ts:10]: Union types provide a more precise type checking for the developers. It's usually better when the subtypes are enumerable.

For `Gradient` I now prefer the interface to be:

```ts
export interface GradientObjectBase {
    id?: number
    type: string
    colorStops: GradientColorStop[]
    global: boolean
}

export interface LinearGradientObject extends GradientObjectBase { };
export interface RadiaGradientObject extends GradientObjectBase { };
export type GradientObject = LinearGradientObject | RadiaGradientObject;
```

File: src/svg/helper/PatternManager.ts [hunk 2]
```diff
@@ -122,7 +122,7 @@ export default class PatternManager extends Definable {
      * @param pattern zr pattern instance
      * @param patternDom DOM to update
      */
-    updateDom(pattern: PatternObject, patternDom: SVGElement) {
+    updateDom(pattern: SVGPatternObject, patternDom: SVGElement) {
```
Comment (pissang) [inline src/svg/helper/PatternManager.ts:125]: Sorry for the confusing name `CanvasPatterObject`. The difference between these two pattern object is whether to use an image in the pattern or use an SVG element. Both of them are supported in SVG renderer. https://github.com/ecomfe/zrender/blob/d30bcf75748ffb40cec43794fddcf89c88660258/src/svg/helper/PatternManager.ts#L139

So perhaps using `ImagePatternObject` and `SVGPatternObject` is clearer.

Comment (plainheart) [reply src/svg/helper/PatternManager.ts:125]: Not sure if 3c2fb9a understands what you mean.

Comment (pissang) [reply src/svg/helper/PatternManager.ts:125]: And SVGPatternObject should not have `image` property.

**Review by pissang — APPROVED** [review]

**Pull request merged**